### DISCUSSION TITLE
Invoke Suspense API

### DISCUSSION THREAD

**Discussion by CreativeTechGuy** [solidjs/solid#1869] [Ideas]
This is an extension of #451.

Currently, there is no way to trigger suspense aside from `createResource`. This means that if we are doing something custom that doesn't fit in `createResource`, the only way to invoke suspense is to add an unnecessary/fake call to `createResource` just for suspense-purposes despite it not actually being needed. This feels wrong and has a lot of unnecessary overhead.

It'd be great if the increment/decrement functions for the internal suspense context are exposed somehow so we can make our own integrations. All of the wiring already exists internally, it could be as simple as exporting `SuspenseContext`.

Example: In my situation, all network traffic is done over a single WebSocket connection. A single "request" to the server may result in multiple responses. As a result, this doesn't work with the `createResource` model. 

TLDR: Can `SuspenseContext` be exported?

**Comment by snnsnn** [comment]
createResource works with any async function, you don't need to make fake calls, it is just a function, does not require you to use fetch or anything. Suspense is intended to be used to coordinate loading states. Your use case may not require direct interaction with SuspenseContext, calling the async function triggers it anyway.

**Reply by CreativeTechGuy** [reply]
By "fake calls", I meant using `createResource` without actually fetching data like it was intended. So we are on the same page. 😃  It feels wasteful to call `createResource` and then not use the return value. I know it'd work, just seems wrong and an anti-pattern. 

@ryansolid recently said (https://github.com/solidjs/solid/issues/1862#issuecomment-1697761104) "the reason for the design current was to prevent need to introduce more signals". That makes me think that signals are expensive in some way. One instance of `createResource` will create 4 signals, let alone do a lot of other stuff, none of which I actually will ever use since I'd only care about the Suspense side effect. Of the roughly ~150 lines of code for `createResource`, the only 2 I actually want to run are the `c.increment()` and `c.decrement()` lines. Everything else is just overhead for my use-case.

If the intended pattern is to do the following, can it be documented somewhere so it is clear that it is intended rather than a hack?

```js
let resolvePromise;
// Create a promise that does nothing to pass to createResource
const promise = new Promise((resolve) => {
    // Store the resolve function so you can resolve the promise due to some external event
    resolvePromise = resolve;
});
// Used only because it enables triggering suspense while the promise is pending
const [data] = createResource(promise);
// Trigger suspense
data();
// Sometime later call: resolvePromise() to end suspense
```

**Reply by snnsnn** [reply]
Yeah, the state of the suspense component is encoded into individual signals, rather than a single object. However signals are cheap, they don't bring that much overhead, but the whole point of resource API is fetching data, fetching data in advance, high up in the scope chain, before it is needed. It may be confusing but you need to be familiar with asynchronous data rendering paradigm. You can start  fetching data when component loads, or you can fetch it in advance long before component is rendered. Resource API allow us do the latter one.

Since you don't need data, and you don't make fetch calls, you may not need suspense and resource API at all, check out this discussion to see how to encode the state of an asynchronous action and device your own API, it will lighter and you will have more control over its execution. https://github.com/solidjs/solid/discussions/1009

> Of the roughly ~150 lines of code for createResource, the only 2 I actually want to run are the c.increment() and c.decrement() lines. Everything else is just overhead for my use-case.

Resource API, Suspense, SuspenseList, Transitions are all parts of a larger API, they are tightly coupled, and they all interact with each other and they interact with the reactive system. They are deeply integrated into the reactive system. The implications of using resource API goes far beyond simple counter, the counter is employed like a tracking mechanism, when a pending resource is read it will increment, when the resource resolve it will decrement, a way to track the overall state to trigger suspense boundary. Changing API will have far reaching consequences than incrementing a value. It will affect even the solid router, etc.

I am not saying it should not be exposed, all i am saying is it will be hard to manage once you introduce another factor into the equation.
